src/shims/ffi_support.rs

```diff
@@ -0,0 +1,265 @@
+                    return Ok(());
+                }
+                // Functions with no declared return type (i.e., the default return)
+                // have the output_type `Tuple([])`.
+                TyKind::Tuple(t_list) =>
+                    if t_list.len() == 0 {
+                        call::<()>(ptr, libffi_args.as_slice());
+                        return Ok(());
+                    },
+                _ => {}
+            }
+            // TODO ellen! deal with all the other return types
+            throw_unsup_format!("unsupported return type to external C function: {:?}", link_name);
+        }
+    }
+
+    /// Call specified external C function, with supplied arguments.
+    /// Need to convert all the arguments from their hir representations to
+    /// a form compatible with C (through `libffi` call).
+    /// Then, convert return from the C call into a corresponding form that
+    /// can be stored in Miri internal memory.
+    fn call_and_add_external_c_fct_to_context(
+        &mut self,
+        link_name: Symbol,
+        dest: &PlaceTy<'tcx, Tag>,
+        args: &[OpTy<'tcx, Tag>],
+    ) -> InterpResult<'tcx, bool> {
+        let this = self.eval_context_mut();
+        let (lib, lib_path) = this.machine.external_so_lib.as_ref().unwrap();
+
+        // Load the C function from the library.
+        // Because this is getting the C function from the shared object file
+        // it is not necessarily a sound operation, but there is no way around
+        // this and we've checked as much as we can.
+        let func: libloading::Symbol<'_, unsafe extern "C" fn()> = unsafe {
+            match lib.get(link_name.as_str().as_bytes()) {
+                Ok(x) => x,
+                Err(_) => {
+                    // Shared object file does not export this function -- try the shims next.
+                    return Ok(false);
+                }
+            }
+        };
+
+        // FIXME: this is a hack!
+        // The `libloading` crate will automatically load system libraries like `libc`.
+        // So, in order to check if the function was actually found in the specified
+        // `machine.external_so_lib` we need to check its `dli_fname` and compare it to
+        // the specified SO file path.
+        // This code is a reimplementation of the mechanism for getting `dli_fname` in `libloading`,
+        // from: https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#411
+        // using the `libc` crate where this interface is public.
+        let mut info = std::mem::MaybeUninit::<libc::Dl_info>::uninit();
+        unsafe {
+            if libc::dladdr(*func.deref() as *const _, info.as_mut_ptr()) != 0 {
+                if std::ffi::CStr::from_ptr(info.assume_init().dli_fname).to_str().unwrap()
+                    != lib_path.to_str().unwrap()
+                {
+                    return Ok(false);
+                }
+            }
+        }
```

Comment:
Can you wrap this together with the `lib.get` above into a function that does the action of "lookup a name in the .so file and make sure it is really from, that file and not one of its dependencies"?